### DISCUSSION TITLE
Possible to hit localhost from database functions and/or Edge functions?

### DISCUSSION THREAD

**Discussion by drewbietron** [supabase/supabase#9837] [Questions]
I am trying to stick to a fully local development setup (locally) and am having troubles making requests to localhost from both a postgres function, or from an Edge Function.  I know that both are ran via the Docker setup from the Supabase CLI, but I've tried about every combination of things I can try in order to make requests to localhost.

I've realized that localhost from the context of a Docker container, is just that container, not the machine's localhost.  As an attempted work around for this, I've tried hitting the exposed Docker IP (in my case `172.25.145.178`) in the database functions and/or Edge functions, but that doesn't work.  I've also added a URL to my hosts config locally so that the IP above can accept requests at a URL vs an IP (eg `app.local -> 172.25.145.178`) which exposes my Nextjs web app and search service to their corresponding ports at app.local, but that doesn't work in the database either.

I plan on storing the url's using Supabase Vault secrets, which will allow me to have different values for different Supabase environments (eg localhost:3000, staging.app.com, production.app.com etc) and reference those variables within my functions or Edge Functions.  I just can't seem to get it wired up to my locally ran services.  Any help would be appreciated.

Here's some example database functions

Tried localhost, that doesnt work since these services aren't ran from within the Supabase local Docker container
```sql
create trigger index_products_to_search
  after insert or update or delete on products
  for each row execute procedure 
  public.send_to_webhook( 'localhost:3000/api/test' );
```

Tried adding a recoerd to hosts to listen at a URL vs localhost but no luck
```sql
create trigger index_products_to_search
  after insert or update or delete on products
  for each row execute procedure 
  public.send_to_webhook( 'http://app.local:3000/api/test' );
```

Would also use the same approach to hit a locally served Edge function.  This one I thought might actually work since this localhost is part of the running Docker container, but no luck.
```sql
create trigger index_products_to_search
  after insert or update or delete on products
  for each row execute procedure 
  public.send_to_webhook( 'http://localhost:54321/functions/v1/function-name' );
```

Any help would be appreciated.  Getting this nailed down would really help me get a great local dev setup going!

**Comment by sweatybridge** [comment]
Hi @drewbietron, you can connect to localhost from within the db container using ip address `172.17.0.1`. For example,

```sql
create trigger index_products_to_search
  after insert or update or delete on products
  for each row execute procedure 
  public.send_to_webhook( 'http://172.17.0.1:54321/functions/v1/function-name' );
```

I haven't tested this trigger per se, so let me know if it works.

**Reply by drewbietron** [reply]
That works!  Thanks @sweatybridge.  Is that IP of the container spun up from Supabase?  How did you find that specific IP?

Is there any way to hit localhost outside of what is in the Docker container?  eg a NextJS app running on port 3000 on the machine vs Docker?

**Reply by sweatybridge** [reply]
Good to know!

> Is that IP of the container spun up from Supabase? How did you find that specific IP?

`172.17.0.1` is the gateway address of docker's [default bridge network](https://docs.docker.com/network/network-tutorial-standalone/). Any container that maps a port to localhost machine is accessible from this ip address within the container.

> Is there any way to hit localhost outside of what is in the Docker container?

I'm not aware of an easy way to do that. One possible alternative to start the NextJS app in a container and join the docker network created by CLI. You can find the network name with `docker network ls`.

**Reply by cayblood** [reply]
@sweatybridge I'm getting a connection refused error when trying to access a web service running on my host OS from a local supabase instance. Here are my docker networks:
```
∴ docker network ls
NETWORK ID     NAME                          DRIVER    SCOPE
2bca90f56cc2   bridge                        bridge    local
e6d11ebccb05   host                          host      local
b1eba64dfec3   none                          null      local
222974e36faa   supabase_network_basejump     bridge    local
ac7a17fb8463   supabase_network_db           bridge    local
314cd019a5ff   supabase_network_finopticon   bridge    local
∴ docker network inspect bridge -f "{{ (index .IPAM.Config 0).Gateway }}"
172.17.0.1
∴ docker network inspect supabase_network_basejump -f "{{ (index .IPAM.Config 0).Gateway }}"
192.168.32.1
∴ docker network inspect supabase_network_db -f "{{ (index .IPAM.Config 0).Gateway }}"
172.18.0.1
∴ docker network inspect supabase_network_finopticon -f "{{ (index .IPAM.Config 0).Gateway }}"
172.20.0.1
```
All of the IPs listed are giving me a connection refused error, even though I can see my web service running and can call it without incident using `curl`. Any suggestions or ideas to try?

**Reply by cayblood** [reply]
FYI I was able to get it working by using `host.docker.internal` instead of `172.17.0.1`.

**Reply by pcnc** [reply]
@cayblood indeed, `host.docker.internal` is the canonical way of reaching the Docker host. 

Unfortunately this is often inconsistent across versions and operating systems as either internal DNS records do not map the host to `host.docker.internal`, or the gateway isn't statically allocated `172.17.0.1`. One of those options will usually work, though.

**Reply by cayblood** [reply]
@pcnc I'm having trouble with this inconsistency right now. I'm developing locally on macOS but running my CI/CD pipeline on ubuntu. My db tests connect to an API mocking service running on localhost. Any suggestions on how I can find a consistent way to connect to the host OS from my local supabase db?

**Reply by cayblood** [reply]
@pcnc for example, it seems like it would be possible to update supabase's docker config to add `host.docker.internal`:

https://stackoverflow.com/questions/31324981/how-to-access-host-port-from-docker-container/61424570#61424570

This would be hugely helpful for me. If you have any influence, I would appreciate any help you could provide to get this implemented. The workarounds are pretty challenging for me. I'm not able to successfully connect on either `host.docker.internal` or `172.17.0.1` on ubuntu, so I'm having to create a publicly-accessible web service for my mock API.

**Reply by pcnc** [reply]
@cayblood just checked across my MacOS and Ubuntu machines and I can confirm, this really is inconsistent.
We'll take a look at it and hope to quickly find a solution addressing this.
Temporarily, you should be able to find the Docker host IP by running `ip -f inet addr show docker0 | awk '/inet / {print $2}'` on the host itself - that is, if the interface naming is also consistent and named `docker0`.

@sweatybridge re [this](https://stackoverflow.com/questions/31324981/how-to-access-host-port-from-docker-container/61424570#61424570) - is it something we could manage when launching containers through the CLI?

**Reply by cayblood** [reply]
@pcnc Running that returns `172.17.0.1/16`, but I'm still getting a connection refused error from my db function when using that IP. If I attempt to connect to the same service running locally on the host using curl to localhost, it works. As I mentioned earlier, connecting to `host.docker.internal` from my db function running locally on my macOS workstation also works.

**Reply by sweatybridge** [reply]
Yup, I have created a PR for adding this extra host config in CLI https://github.com/supabase/cli/pull/821. Aiming to release over the next day or two.

**Reply by cayblood** [reply]
Thanks so much @sweatybridge and @pcnc! Love the responsiveness here.

**Reply by bombillazo** [reply]
Hey, can we use `http://localhost:X` as the endpoint? We have a URL parameter for things like redirects and we'd like to keep the consistency in logic for all environments.

**Reply by rbkayz** [reply]
@sweatybridge -  this trigger is in my migration SQL. Any clue on how to customize the first part of the URL such that when we deploy the migration

http://172.17.0.1:54321 -> https://<project-id>.supabase.co

**Reply by samuelbohl** [reply]
Hi @sweatybridge I also here to ask about the issue @rbkayz mentioned.

The question is: How can I setup a trigger so that it calls a supabase edge function in the correct environment, without having to somehow split the migration scripts. Is there a solution that supabase provides here?

**Reply by rbkayz** [reply]
Wrote a function for this. The only thing you'll need to do is add the project_id in supabase vault

Here's the code

```
CREATE OR REPLACE FUNCTION webhooks_func()
	RETURNS TRIGGER
	SECURITY DEFINER
	LANGUAGE plpgsql
	AS $$
DECLARE
	url text;
	function_name text := TG_ARGV[0]::text;
	project_id text;
	project_env text DEFAULT 'LOCAL';
	payload jsonb;
	headers jsonb;
	request_id bigint;
	timeout_ms integer DEFAULT 1000;
BEGIN
	--
	--
	IF function_name IS NULL THEN
		RAISE EXCEPTION 'function_name cannot be null';
	END IF;
	--
	-- GET URL ENDPOINT
	SELECT
		decrypted_secret
	FROM
		vault.decrypted_secrets
	WHERE
		name = 'project_id'
	LIMIT 1 INTO project_id;
	--
	-- GET ENV
	SELECT
		decrypted_secret
	FROM
		vault.decrypted_secrets
	WHERE
		name = 'project_env'
	LIMIT 1 INTO project_env;
	--
	--
	IF project_id IS NULL THEN
		url := 'http://172.17.0.1:54321';
	ELSE
		url := 'https://' || project_id || '.supabase.co';
	END IF;
	--
	--
	IF project_env IS NULL THEN
		project_env := 'LOCAL';
	END IF;
	--
	--
	headers := '{"Content-type":"application/json"}'::jsonb;
	--
	--
	payload := jsonb_build_object('old_record', OLD, 'record', NEW, 'type', TG_OP, 'table', TG_TABLE_NAME, 'schema', TG_TABLE_SCHEMA, 'env', project_env);
	--
	--
	SELECT
		http_post INTO request_id
	FROM
		net.http_post(url || '/functions/v1/' || function_name, payload, '{}'::jsonb, headers, timeout_ms);
	--
	--
	INSERT INTO supabase_functions.hooks(
		hook_table_id,
		hook_name,
		request_id)
	VALUES (
		TG_RELID,
		TG_NAME,
		request_id);
	--
	--
	RETURN NEW;
END
$$;
```

Then create the trigger as follows with the name of your function

```
CREATE OR REPLACE TRIGGER webhook_after_insert_tbl_org_members
	AFTER INSERT ON tbl_org_members
	FOR EACH ROW
	EXECUTE FUNCTION webhooks_func('invite-members');
```

Good part is that this goes in the migration when you diff the public schema (as opposed to setting the trigger from the dashboard UI)

**Reply by rbkayz** [reply]
@samuelbohl

**Reply by samuelbohl** [reply]
Thanks! @rbkayz The authorization header is missing, but I guess I can also just store it in the vault?

**Reply by rbkayz** [reply]
You can. Or you can pass the `--no-verify-jwt` flag when deploying these trigger functions

**Comment by ChuckJonas** [comment]
This should really be mentioned [somewhere in the docs](https://supabase.com/docs/guides/functions/local-development).  It took me way too long to track this down

**Reply by tomtitherington** [reply]
+1 I spent an entire morning wondering why it wasn't working...

**Comment by fabrizioanichini** [comment]
Is there any easy way right now?

**Comment by floross** [comment]
Hi, I am in macos and I am still not able to call my local nestjs api serveur (running on 3000) from a database trigger. In the net._http_response I still got `Couldn't connect to server` error. I tried to use `http://host.docker.internal:3000/api` and `http://172.17.0.1:3000/api` without any success. I used ngrok now but I'd prefered to stay on my local computer (I don't have internet all the time).

Is there another way to hit my local api from the supabase cli database ?

**Reply by jrpersico** [reply]
Did you find a solution?

**Reply by chang-ryan** [reply]
curious here too, i'm getting the same problems when trying to use `host.docker.internal`

**Reply by chang-ryan** [reply]
<img width="751" alt="Screenshot 2024-03-01 at 4 01 40 PM" src="https://github.com/supabase/supabase/assets/13877301/1e57e1bc-9e59-4996-8105-3babf1ba3bd8">

**Reply by chang-ryan** [reply]
not working for me

**Reply by simon-inkie** [reply]
I had this issue, many hours spent. In the end it was my vite config, you need to allow all hosts (it defaults to localhost only)
`  server: {
    port: 3000,
    host: '0.0.0.0'
  },`
https://vite.dev/config/server-options

**Comment by devjume** [comment]
Created following snippet that creates trigger to call edge function endpoint based on environment.

```
DO $$
DECLARE
    function_name TEXT := 'hello_world'; 
    trigger_name TEXT := 'new_test_run_hello_world';
    table_name TEXT := 'public.test';
    production_url TEXT := 'https://XXXXX.supabase.co'; 
    development_url TEXT := 'http://host.docker.internal:54321'; 
    function_url TEXT := ''; -- Dont edit. This will be set dynamically
    server_ip INET;
BEGIN
    --- DONT EDIT BELOW
    SELECT inet_server_addr() INTO server_ip;

    function_url := 
    CASE 
        WHEN server_ip >= '172.0.0.0'::inet AND server_ip <= '172.255.255.255'::inet
        THEN development_url || '/functions/v1/' || function_name
        ELSE production_url || '/functions/v1/' || function_name
    END;
    --- DONT EDIT UP
	
    -- FOR DEBUGGING
    -- RAISE EXCEPTION 'Test %, %, %', production_url, development_url, function_url;

     -- EDIT FOLLOWING TO SETUP YOUR TRIGGER

    EXECUTE format('
        CREATE OR REPLACE TRIGGER %s
            AFTER INSERT
            ON %s
            FOR EACH ROW
            EXECUTE FUNCTION supabase_functions.http_request(''%s'', ''POST'', ''{"Content-type":"application/json"}'', ''{"name":"Person1"}'', ''1000'');
    ', trigger_name, table_name, function_url);

END $$;

```

Pros: 

- It checks if postgres database IP is in 172.0.0.0-172.255.255.255 range (meaning local docker), and sets the webhook url based on that to localhost or production url.
- Webhooks will show in the UI. Database > Webhooks.
- Can be used with migrations and version controlled.

Cons:

- Each webhook must be added manually to the migration file by using the snippet.
- Supabase project url is exposed to git repository via the migration files. But I treat that as public information anyway.

I use the snippet in following development workflow when I want to trigger edge function on database event. In bot local development and production.

1. Create new edge function: `supabase functions new HELLO_WORLD`
2. Code the edge function
3. Server edge function locally `supabase functions serve`
4. Add it to local Supabase instance using in SQL editor **by running previous snippet**
5. Debug and verify that everything works
6. Create new db migration file `supabase migration new DB_TRIGGER_WEBHOOK_XXX` 
7. Test migrations locally and verify again that everything works `supabase db reset`
8. Deploy edge function to production `supabase functions deploy hello-world`
9. Deploy migration to production `supabase db push`
10. Profit

I didn’t fully test this with supabase migration as I use drizzle. But I am pretty sure this works as stated.

Edit: Better IP check and instread localhost it uses `host.docker.internal`

**Comment by RyanClementsHax** [comment]
For those that don't want to use a hardcoded ip like `172.17.0.1`, you can take advantage of docker's networking and use the `db` alias.

Ex:
```
postgresql://authenticator:postgres@db:5432/postgres
```

**Comment by magick93** [comment]
Did anyone find a way to achieve this using Linux?

**Comment by kkukshtel** [comment]
It took me so long to find this answer and thank god for the thread there's no way I would have thought to check Docker. It should be emitted as part of `supabase start` imo.

`http://172.17.0.1:54321` worked for me.

**Comment by cwooldridge1** [comment]
for local host from edge functions you can use `host.docker.internal`

e.g. 
```
    const response = await fetch(`http://host.docker.internal:3000/api`, {
      method: "GET",
      headers: {
        "Content-Type": "application/json",
      },
    });
```

**Comment by Auuufff** [comment]
If someone is struggling with "_Couldn't connect to server_" or "_Timeout was reached_" on self-host docker supabase, you can try to use that URL for your webhook `http://kong:8000/functions/v1/your-function`

**Reply by GitTorres** [reply]
This was the only URL that worked for me in May 2025 that allowed my edge function to use the supabase api client to query data.

**Comment by chidg** [comment]
The thread is going back a while but I had this problem today and perhaps someone will find this helpful. In order to allow a database trigger function to call an edge function, you can use this url in development `http://api.supabase.internal:8000`, and in production, the project's API url. This is documented in this blog post https://supabase.com/docs/guides/ai/automatic-embeddings.

Hope that helps, it works for me.

**Reply by tarcanarca** [reply]
This solution worked for me. It eliminates the need for hardcoded project URLs or logic to determine them in code, just inject them via secrets.

**Comment by aantti** [comment]
Somewhat related - #37271, leaving it here for visibility.

**Comment by Rubenburdin** [comment]
From database triggers to local services: use http://172.17.0.1:YOUR_PORT on Linux or http://host.docker.internal:YOUR_PORT on Mac/Windows. For calling local Edge Functions specifically, http://api.supabase.internal:8000/functions/v1/your-function works well.
From Edge Functions to local services: same thing, use host.docker.internal in your fetch calls.
Common gotcha: your local dev server (Vite, Next, etc) probably binds to 127.0.0.1 by default. Docker can't reach that. Bind to 0.0.0.0 instead:
```
// Vite
server: { host: '0.0.0.0' }

// Next  
next dev -H 0.0.0.0
```
That's usually why "Couldn't connect to server" happens even with the right Docker URL.

**Comment by aantti** [comment]
@kallebysantos jfyi

**Comment by aantti** [comment]
I've added a note on internal vs external URLs here: https://supabase.com/docs/guides/self-hosting/self-hosted-functions#internal-vs-external-urls (as well as adding `SUPABASE_PUBLIC_URL` to the configuration of `functions` in [self-hosted Supabase](https://supabase.com/docs/guides/self-hosting))
